runtime/vm/FastJNI_com_ibm_crypto_plus_provider_icc_NativeInterface.cpp

```diff
@@ -0,0 +1,415 @@
+/*******************************************************************************
+ * Copyright (c) 2001, 2020 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#include "fastJNI.h"
+
+#include "j9protos.h"
+#include "j9consts.h"
+#include "objhelp.h"
+#include "BytecodeAction.hpp"
+#include "VMHelpers.hpp"
+#include "ArrayCopyHelpers.hpp"
+
+//#define CRYPTO_DEBUG 0
+
+extern "C" {
+
+typedef int CIPHER_t(long iccContextId, long iccCipherId, void* input, int inputLen, void* output, int needsReinit);
+typedef int z_kmc_native_t(void* input, void* output, int inputLength, long paramPointer, int mode);
+
+typedef union {
+	CIPHER_t* CIPHER_op;
+	z_kmc_native_t* z_kmc_native;
+} cryptoFunc_t;
+
+typedef struct cryptoPointers {
+	cryptoFunc_t CIPHER_encryptFinalPtr;
+	cryptoFunc_t CIPHER_decryptFinalPtr;
+	cryptoFunc_t CIPHER_encryptUpdatePtr;
+	cryptoFunc_t CIPHER_decryptUpdatePtr;
+	cryptoFunc_t z_kmc_nativePtr;
+	
+	CIPHER_t* CIPHER_encryptUpdate;
+	CIPHER_t* CIPHER_decryptUpdate;
+} cryptoPointers_t;
+
+
+/* loads crypto lib routines and store routine pointers for use.
+ * Need to synchronize this operation so that threads won't
+ * use the routines too early
+ * */
+void loadCryptoLib(J9VMThread *currentThread) {
+
+	if (currentThread->javaVM->jniCryptoFunctions == NULL) {
+
+		omrthread_monitor_t mutex = currentThread->javaVM->jniCryptoLibLock;
+		omrthread_monitor_enter(mutex);
+		
+		if (currentThread->javaVM->jniCryptoFunctions == NULL) {
+	
+			PORT_ACCESS_FROM_VMC(currentThread);
+			cryptoPointers_t* cryptFuncs = (cryptoPointers_t*)j9mem_allocate_memory(sizeof(cryptoPointers_t), J9MEM_CATEGORY_VM_JCL);
+			UDATA* cryptoLibrary = &(currentThread->javaVM->jniCryptoLibrary);
+	
+			CIPHER_t* CIPHER_encryptFinal;
+			CIPHER_t* CIPHER_decryptFinal;
+			z_kmc_native_t* z_kmc_native;
+			const char * cryptolib_name = "jgskit";
```

Comment:
I don't see the point of declaring these rather than just using them directly in the port library calls.

Reply:
without declaring those. For example, if the code was

``j9sl_open_shared_library("jgskit", cryptoLibrary, J9PORT_SLOPEN_DECORATE);``

The compiler will generate an error such as:

```FastJNI_com_ibm_crypto_plus_provider_icc_NativeInterface.cpp: In function 'void loadCryptoLib(J9VMThread*)':
../oti/j9port_generated.h:553:180: error: ISO C++ forbids converting a string constant to 'char*' [-Werror=write-strings]
 #define j9sl_open_shared_library(param1,param2,param3) OMRPORT_FROM_J9PORT(privatePortLibrary)->sl_open_shared_library(OMRPORT_FROM_J9PORT(privatePortLibrary),param1,param2,param3)
```

Reply:
Fair enough - the correct solution would be to consitfy the port library, but that's not going to happen here.